clippy_lints/src/matches.rs

```diff
@@ -1602,17 +1759,36 @@ mod redundant_pattern_match {
                 }
             },
             PatKind::Path(ref path) => {
-                if match_qpath(path, &paths::OPTION_NONE) {
+                let method = if match_qpath(path, &paths::OPTION_NONE) {
                     "is_none()"
                 } else if match_qpath(path, &paths::POLL_PENDING) {
                     "is_pending()"
                 } else {
                     return;
-                }
+                };
+                // `None` and `Pending` don't have an inner type.
+                (method, cx.tcx.types.unit)
             },
             _ => return,
         };
 
+        // If this is the last expression in a block or there is an else clause then the whole
+        // type needs to be considered, not just the inner type of the branch being matched on.
+        // Note the last expression in a block is dropped after all local bindings.
+        let check_ty = if has_else
+            || (keyword == "if" && matches!(cx.tcx.hir().parent_iter(expr.hir_id).next(), Some((_, Node::Block(..)))))
+        {
+            op_ty
+        } else {
+            inner_ty
+        };
+
+        // All temporaries created in the scrutinee expression are dropped at the same time as the
+        // scrutinee would be, so they have to be considered as well.
+        // e.g. in `if let Some(x) = foo.lock().unwrap().baz.as_ref() { .. }` the lock will be held
+        // for the duration if body.
+        let needs_drop = type_needs_ordered_drop(cx, check_ty) || temporaries_need_ordered_drop(cx, op);
```

Comment:
This check **has** to be somewhere in MIR already.  IIUC This is basically checking for this: https://doc.rust-lang.org/stable/reference/destructors.html#temporary-scopes

Am I correct, that you want to check for this behavior?

Reply:
That is what I'm checking for. There should be some way, but I don't know how to link MIR to the expression it came from. It would be the better solution.

Reply:
I also am not experienced with the MIR. @oli-obk can you help out here, please?

Reply:
The MIR indeed contains this information. But in order to leverage that information, the lint would need to be rewritten as a MIR lint, as it is nigh impossible to map between the HIR and the MIR. In the MIR all `Drop` invocations are explicit, so similarly to our "unnecessary clone lint", we can write a dataflow analysis on MIR that will tell us at any specific location in the MIR whether a specific local is going to be dropped by the time we reach that location.

Rewriting as a MIR lint should be done independently from any changes, as to not accidentally introduce changes.

Reply:
Given that the lint operates specifically on match expressions, and that we'd be looking specifically at the temporary values created in the scrutinee expression, is it fair to say that would be non-trivial to look for in the MIR?

Is there any way to use the code lowering the HIR into the MIR to find if there are any temporary values created by an expression which will need to subsequently be dropped?

Reply:
@oli-obk can I get you to weigh in here one more time, please?

Reply:
I don't see a good way to do this on the MIR at this time. The information is indeed available during HIR -> MIR lowering (well... in the THIR, which is between the two). The THIR is being heavily refactored at the moment and unsafety checking will be built on THIR. It may be reasonable to implement this analysis on THIR at that point, but not now.

So...

> is it fair to say that would be non-trivial to look for in the MIR?

yes

> Is there any way to use the code lowering the HIR into the MIR to find if there are any temporary values created by an expression which will need to subsequently be dropped?

Not yet. So proceeding with the PR as it is written seems good

Reply:
Thank you.